src/main/java/com/google/devtools/build/lib/syntax/MethodLibrary.java

```diff
@@ -170,14 +186,45 @@ private static boolean hasElementWithBooleanValue(
             type = Object.class,
             doc = "This collection.",
             // TODO(cparsons): This parameter should be positional-only.
-            legacyNamed = true)
+            legacyNamed = true),
+        @Param(
+            name = "key",
+            doc = "The key to sort with.",
+            named = true,
+            defaultValue = "None",
+            positional = false,
+            noneable = true),
+        @Param(
+            name = "reverse",
+            type = Boolean.class,
+            doc = "Whether to sort asc or desc.",
+            named = true,
+            defaultValue = "False",
+            positional = false,
+            noneable = true)
       },
       useLocation = true,
       useEnvironment = true)
-  public MutableList<?> sorted(Object self, Location loc, Environment env) throws EvalException {
+  public MutableList<?> sorted(Object self, Object key, Boolean reverse,
+      Location loc, Environment env) throws EvalException, InterruptedException {
+    Iterator<?> iterator = EvalUtils.toCollection(self, loc, env).iterator();
+    List<Object> list = new ArrayList<>();
+
+    while (iterator.hasNext()) {
+      Object val = iterator.next();
+      list.add(new EvalUtils.SortPair(evalKey(val, key, loc, env), val));
+    }
+
     try {
-      return MutableList.copyOf(
-          env, EvalUtils.SKYLARK_COMPARATOR.sortedCopy(EvalUtils.toCollection(self, loc, env)));
+      list = EvalUtils.SKYLARK_COMPARATOR.sortedCopy(list);
```

Comment:
Use Collections.sort passing an explicit comparator. The comparator will apply the key function and then invoke the SKYLARK_COMPARATOR.

Reply:
The comparator's compare function isn't designed to throw a checked exception (InterruptedException and EvalException) which the key function would throw if applied from within. I could still mask these exceptions inside a RuntimeException but I'm not sure if that's a better solution than the current one.

Reply:
You can either create a comparator that catches and retains any exception thrown (like the Go implementation does), or you can wrap the checked exception and throw it through the JDK sort package. The 'sorted' function would then inspect, unwrap, and rethrow any error.

Either of those solutions is preferable to exposing a new data type in the public API.